Support passing a connection URL to Trino CLI

nineinchnick opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

Allow the Trino CLI to accept an URL parameter instead of a (sometimes long) list of options. URLs are parsed same as JDBC URLs, but they don't have the `jdbc:` prefix. If any parameters are set in the URL, they can't be also set as options. But they can be present in the config file because these are only treated as default values and can be overridden.

> Is this change a fix, improvement, new feature, refactoring, or other?
new feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
cli

> How would you describe this change to a non-technical end user or system administrator?
Trino CLI can accept an URL parameter.

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->
* Fixes #12543

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# CLI
* Trino CLI now accepts a single URL with parameters as an alternative to passing command line arguments. ({issue}`12587`)

# JDBC
* Connection parameter validation errors are now actionable ({issue}`12587`)

```

nineinchnick commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -113,23 +166,23 @@
     @Option(names = "--external-authentication-redirect-handler", paramLabel = "<externalAuthenticationRedirectHandler>", description = "External authentication redirect handlers: ${COMPLETION-CANDIDATES} " + DEFAULT_VALUE, defaultValue = "ALL")
     public List<ExternalRedirectStrategy> externalAuthenticationRedirectHandler = new ArrayList<>();
 
-    @Option(names = "--source", paramLabel = "<source>", defaultValue = "trino-cli", description = "Name of source making query " + DEFAULT_VALUE)
```
I should revert removing the defaultValue here, as I later figured out how to identify which options are set by the user.

nineinchnick reviewed (commented):
To the reviewers, I understand this is pretty big and I'll try to split it up into multiple smaller commits.

nineinchnick commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1196 @@
+        }
+
+        Properties result = new Properties();
+        setProperties(result, suppliedProperties);
```
Make sure there's a test to check that properties take precedence over URL parameters.

nineinchnick commented on client/trino-cli/src/main/java/io/trino/cli/Console.java:
```diff
@@ -104,7 +108,16 @@ public Integer call()
+                .map(CommandLine.Model.OptionSpec::longestName)
+                .collect(Collectors.toList());
+
+        TrinoUri uri = clientOptions.getTrinoUri(providedOptions);
```
This has a nice side effect of always converting all CLI options into URLs that _can_ be used as JDBC URLs. Would it be useful to log the resulting connection URL on a debug level?

MiguelWeezardo replied:
If it won't expose the password I'm for it.

nineinchnick commented:
> To the reviewers, I understand this is pretty big and I'll try to split it up into multiple smaller commits.

Done. If there's anything else I can do to make this easier to review please let me know.

mosabua commented on docs/src/main/sphinx/installation/cli.rst:
```diff
@@ -188,6 +188,12 @@ mode:
       operating system username. You can override the default username,
       if your cluster uses a different username or authentication mechanism.
 
+Most of the options can also be set as parameters in the URL. This means
```
Most options can

mosabua commented on docs/src/main/sphinx/installation/cli.rst:
```diff
@@ -188,6 +188,12 @@ mode:
+a JDBC URL can be used in the CLI after removing the ``jdbc:`` prefix.
+However, the same parameter may not be specified using both methods.
+See :doc:`the JDBC driver parameter reference </installation/jdbc>`
+to find out URL parameter names.
```
this will need an example or two..

nineinchnick replied:
Added. I also added a note saying that the password parameter cannot be set in the URL, PTAL.

mosabua reviewed (changes requested):
Docs only need minor changes. From the behaviour side of things I am not sure we want the new behavior to be the default and hence the one we use in the documentation. 

And importantly I would like confirmation that the old behavior will continue to work...

nineinchnick commented:
> Docs only need minor changes. From the behaviour side of things I am not sure we want the new behavior to be the default and hence the one we use in the documentation.

Let's wait for others to chime in about this. I made it the default since it's shorter. Users won't have to think about the option name (was it `server`, `url`, `address` or `host`?) and generally should be easier to type.

> And importantly I would like confirmation that the old behavior will continue to work... and we will continue to ensure that with test coverage.

Yes, no existing behavior should change and there are lots of tests for that. I added a few test cases with URLs.

willmostly commented:
Useful feature for debugging @nineinchnick, looking forward to this!

mosabua commented:
> > Docs only need minor changes. From the behaviour side of things I am not sure we want the new behavior to be the default and hence the one we use in the documentation.
> 
> Let's wait for others to chime in about this. I made it the default since it's shorter. Users won't have to think about the option name (was it `server`, `url`, `address` or `host`?) and generally should be easier to type.

My concern is that it is NOT easier to type for any real use case beyond the URL only. In fact you need to know how to format the JDBC URL string. 

My other questions would be .. how much does this work if you combine JDBC string and other parameters. Are all parameters from both input methods combined somehow? What if there are collisions .. does on override the other? Which one wins out? We need to make sure we understand that and also document it.

nineinchnick commented:
> My concern is that it is NOT easier to type for any real use case beyond the URL only. In fact you need to know how to format the JDBC URL string.

I meant it's easier only in the most basic use case, `trino --server https://some.host --catalog bigdata --schema foo` vs `trino https://some.host/bigdata/foo`. 

> My other questions would be .. how much does this work if you combine JDBC string and other parameters. Are all parameters from both input methods combined somehow? What if there are collisions .. does on override the other? Which one wins out? We need to make sure we understand that and also document it.

You can combine options, start with the URL only and add more options instead of URL parameters. You can't specify the same parameter both in the URL and as a CLI option. There's one sentence clarifying this in the docs update.

nineinchnick commented:
@electrum PTAL

nineinchnick commented on client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractConnectionProperty.java:
```diff
@@ -113,13 +107,20 @@ public boolean isAllowed(Properties properties)
         if (value == null) {
-            if (isRequired(properties)) {
+            if (isRequired(properties) && !defaultValue.isPresent()) {
                 throw new SQLException(format("Connection property '%s' is required", key));
```
Isn't that exactly the same? Required sounds a bit more explicit and actionable.

nineinchnick commented on client/trino-client/src/main/java/io/trino/client/ClientSession.java:
```diff
@@ -300,8 +300,12 @@ private Builder(ClientSession clientSession)
+                catalog = clientSession.getCatalog().get();
+            }
+            if (clientSession.getSchema().isPresent()) {
+                schema = clientSession.getSchema().get();
```
Because they're not optional in the builder.

nineinchnick commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,632 @@
+                        redirectHandler, poller, knownTokenCache.create(), timeout);
+
+                builder.authenticator(authenticator);
+                builder.addInterceptor(authenticator);
```
I did not write this code, just moved it to this module from the JDBC driver.

wendigo replied:
please extract it to the a separate commit (class move)

nineinchnick commented on client/trino-cli/src/test/java/io/trino/cli/TestClientOptions.java:
```diff
@@ -48,79 +51,115 @@ public void testDefaults()
+        console.clientOptions.toClientSession(console.clientOptions.getTrinoUri());
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class, expectedExceptionsMessageRegExp = "Using both the URL parameter and the --server option is not allowed")
```
Done!

nineinchnick commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -574,18 +791,24 @@ private void initCatalogAndSchema()
         }
 
+        if (catalog.isPresent()) {
+            throw new RestrictedPropertyException(PropertyName.CATALOG, "Catalog cannot be set in the URL");
```
There's another occurrence of this on line 661:
```
            if (restrictedProperties.contains(name)) {
                throw new RestrictedPropertyException(name, format("Connection property '%s' cannot be set in the URL", parts.get(0)));
            }
```

nineinchnick replied:
Ah I just remembered that this is exactly the reason why I added a custom exception class. I catch it in `ClientOptions.java` and rewrite the message into something more actionable:
```java
        catch (RestrictedPropertyException e) {
            if (e.getPropertyName() == PropertyName.PASSWORD) {
                throw new IllegalArgumentException("Setting the password in the URL parameter is not allowed, use the `--password` option or the `TRINO_PASSWORD` environment variable");
            }
            Map<PropertyName, String> propertiesToOptions = optionsToProperties.entrySet()
                    .stream()
                    .collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));
            throw new IllegalArgumentException(format(
                    "Connection property '%s' cannot be set in the URL when option '%s' is set",
                    e.getPropertyName(),
                    propertiesToOptions.get(e.getPropertyName())), e);
        }
```

electrum commented:
We'd eventually like to make the CLI use the JDBC driver, so moving things into the client is a step in the wrong direction. Let's keep `TrinoDriverUri` part of JDBC.

nineinchnick commented:
> We'd eventually like to make the CLI use the JDBC driver, so moving things into the client is a step in the wrong direction. Let's keep `TrinoDriverUri` part of JDBC.

@electrum I started working on reverting this change but I'm running into some issues I need help with. I'd have to import trino-jdbc into trino-cli, but it turns out that trino-jdbc shades trino-client, and it's causing compilation errors where trino-cli tries to pass types from trino-client to methods from trino-jdbc:
```
[ERROR] /Users/jwas/src/trino/client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:[373,47] incompatible types: java.util.List<io.trino.client.auth.external.ExternalRedirectStrategy> cannot be converted to java.util.List<io.trino.jdbc.$internal.c
lient.auth.external.ExternalRedirectStrategy>
```
I don't know how to deal with this.

Going back to what you said about trino-cli using trino-jdbc, does this mean that trino-jdbc and trino-client will be merged at some point?

nineinchnick commented:
@electrum I'd appreciate any suggestion on how to continue here. I can do more research if required.

nineinchnick commented:
@electrum I'd still use some pointers on how to proceed.

nineinchnick commented:
I looked at how much work would be to fully convert to only using the JDBC and such CLI would be very limited with the current JDBC driver. For example, there's no way of reporting the query progress and getting statistics. But this PR is a step forward, as it unifies the way all connection properties are being set.

wendigo commented on client/trino-client/src/main/java/io/trino/client/DnsResolver.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.List;
+
+public interface DnsResolver
+{
```
Does it belong to this commit?

wendigo replied:
If not, please remove

nineinchnick replied:
That's a leftover from a bad rebase, I moved it to the last commit.

wendigo commented on client/trino-jdbc/src/main/java/io/trino/jdbc/PropertyName.java:
```diff
@@ -0,0 +1,95 @@
+
+import static java.lang.String.format;
+
+public enum PropertyName
```
nit: `ConnectionPropertyName` ?

nineinchnick replied:
There's no other type of properties in the jdbc package so I'd prefer the shorter name.

wendigo commented on client/trino-jdbc/src/main/java/io/trino/jdbc/PropertyName.java:
```diff
@@ -0,0 +1,95 @@
+        for (PropertyName name : PropertyName.values()) {
+            lookup.put(name.toString(), name);
+        }
+    }
```
just:
```suggestion
    private static final Map<String, PropertyName> lookup = stream(values())
            .collect(toImmutableMap(Enum::name, identity()));
```

wendigo replied:
bump

nineinchnick replied:
Done

wendigo commented on client/trino-jdbc/src/main/java/io/trino/jdbc/PropertyName.java:
```diff
@@ -0,0 +1,95 @@
+        return key;
+    }
+
+    public static PropertyName get(String key)
```
Instead, let's do make it:
```suggestion
    public static Optional<PropertyName> findByKey(String key)
```
so that method doesn't throw. Let's use if in `TrinoDriverUri` without try catch

wendigo replied:
bump

nineinchnick replied:
I haven't pushed the changes yet, since I want to address all comments first. I only replied to comments that didn't require changes.

nineinchnick replied:
Done

wendigo commented on client/trino-jdbc/src/main/java/io/trino/jdbc/ConnectionProperties.java:
```diff
@@ -540,7 +525,7 @@ public ExternalAuthenticationRedirectHandlers()
             super(
                     PropertyName.EXTERNAL_AUTHENTICATION_REDIRECT_HANDLERS,
-                    Optional.of("OPEN"),
+                    Optional.of(Collections.singletonList(ExternalRedirectStrategy.OPEN)),
```
static import singletonList

wendigo commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -22,35 +22,88 @@
+            .put("--schema", PropertyName.SCHEMA)
+            .buildOrThrow();
+
+    @Parameters(paramLabel = "URL", description = "Trino server URL", arity = "0..1")
```
let's call the label `Connection URI` and `Trino connection URI`

nineinchnick replied:
But it's not a URI, it has to be a valid location.

nineinchnick replied:
Also, why the longer version? It shows up in help as:

```

trino [-h] [--debug] ... [URL]
      [URL]                  Trino server URL
```

wendigo commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -249,10 +302,126 @@ public ClientSession toClientSession()
 
+    public TrinoUri getTrinoUri()
+    {
+        return getTrinoUri(Collections.emptyList());
```
static import `emptyList`

wendigo commented on client/trino-client/src/main/java/io/trino/client/uri/RestrictedPropertyException.java:
```diff
@@ -11,14 +11,23 @@
  * limitations under the License.
  */
-package io.trino.jdbc;
+package io.trino.client.uri;
```
please extract moves like that to a separate commit

wendigo commented on docs/src/main/sphinx/client/cli.rst:
```diff
@@ -102,7 +102,7 @@ allows you to query tables directly without specifying catalog and schema.
 
 .. code-block:: text
 
-  ./trino --server http://trino.example.com:8080 --catalog tpch --schema tiny
```
does the previous `--server` still works?

nineinchnick replied:
Yes, this change is fully compatible. I updated the docs since the new syntax is shorter and easier to type and doesn't require remembering flag names.

wendigo replied:
Ok. I just wanted to make sure that we are not changing the previous behavior.

wendigo commented on client/trino-client/src/main/java/io/trino/client/StatementClientV1.java:
```diff
@@ -149,11 +149,11 @@ private Request buildQueryRequest(ClientSession session, String query)
         }
-        if (session.getSchema() != null) {
-            builder.addHeader(TRINO_HEADERS.requestSchema(), session.getSchema());
+        if (session.getSchema().isPresent()) {
```
```suggestion
        session.getSchema().ifPresent(value -> builder.addHeader(TRINO_HEADERS.requestSchema(), value));
```

wendigo commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -249,10 +302,126 @@ public ClientSession toClientSession()
+        return getTrinoUri(Collections.emptyList());
+    }
+
+    public TrinoUri getTrinoUri(List<String> providedOptions)
```
can we extract this method to a separate class and provide some tests for it?

nineinchnick replied:
There are lots of test for it in `TestClientOptions` already.

wendigo commented on client/trino-cli/src/main/java/io/trino/cli/Console.java:
```diff
@@ -102,7 +106,16 @@ public Integer call()
     public boolean run()
     {
-        ClientSession session = clientOptions.toClientSession();
+        // map option names to URL params and restrict those set explicitly by the user
```
extract this logic to a separate method `private static List<String> getCommandLineArguments(CommandSpec spec)`

wendigo commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -22,35 +22,88 @@
+
+    // some entries have different semantics, but this mapping is only used to disallow setting both as a CLI option and URL parameter
+    static final Map<String, PropertyName> optionsToProperties = ImmutableMap.<String, PropertyName>builder()
+            .put("--krb5-service-principal-pattern", PropertyName.KERBEROS_SERVICE_PRINCIPAL_PATTERN)
```
can we add these bits to the `PropertyName` itself?

```
PropertyName(final String key, final String commandLineKey, boolean allowedInUri)
```

nineinchnick replied:
This is a map of cli options to property names. Since the `PropertyName` class is defined in `trino-client` it shouldn't include cli options. When a user specifies both a URL and an option, we only allow the URL to contain properties that are not set as options, to avoid silently overriding them, which can be confusing. So all properties are allowed by default, unless some options are set by the user.

wendigo replied:
Makes sense but how to ensure that those two lists are in-sync (CLI and PropertyName-s)?

wendigo replied:
Let's introduce an annotation applied to the CLI Options with the PropertyName mapping

nineinchnick replied:
I pulled in your suggestion

wendigo commented on client/trino-cli/src/main/java/io/trino/cli/Console.java:
```diff
@@ -246,9 +246,9 @@ private static void runConsole(
-                String schema = queryRunner.getSession().getSchema();
-                if (schema != null) {
-                    prompt += ":" + schema.replace("%", "%%");
+                Optional<String> schema = queryRunner.getSession().getSchema();
+                if (schema.isPresent()) {
+                    prompt += ":" + schema.get().replace("%", "%%");
```
Nit: this can be simplified as (no change requested):
```suggestion
                prompt += schema.map(value -> ":" + value.replace("%", "%%"))
                		.orElse("");
```

nineinchnick replied:
Done

wendigo commented on client/trino-client/src/main/java/io/trino/client/ClientSession.java:
```diff
@@ -307,8 +307,12 @@ private Builder(ClientSession clientSession)
-            catalog = clientSession.getCatalog();
-            schema = clientSession.getSchema();
+            if (clientSession.getCatalog().isPresent()) {
+                catalog = clientSession.getCatalog().get();
```
can we just assign `catalog = clientSession.getCatalog().orElse(null)`, the same for schema? latter in the code we are converting that to the `Optional` with the `.ofNullable` call

nineinchnick replied:
Done

mosabua commented on docs/src/main/sphinx/client/cli.rst:
```diff
@@ -56,7 +56,7 @@ the coordinator in the Trino cluster:
 .. code-block:: text
 
-    ./trino --server http://trino.example.com:8080
+    ./trino http://trino.example.com:8080
```
If  --server becomes optional we should update the docs for that,

mosabua commented on docs/src/main/sphinx/client/cli.rst:
```diff
@@ -206,6 +206,12 @@ mode:
+Most of the options can also be set as parameters in the URL. This means
+a JDBC URL can be used in the CLI after removing the ``jdbc:`` prefix.
+However, the same parameter may not be specified using both methods.
+See :doc:`the JDBC driver parameter reference </client/jdbc>`
```
We should have at least one example of a jdbc url and show it here explicitly,

nineinchnick replied:
Added, PTAL.

wendigo commented:
@mosabua PTAL for doc part

mosabua reviewed (approved):
Docs look good now.

nineinchnick commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriverUri.java:
```diff
@@ -89,8 +89,8 @@ public void testInvalidUrls()
-        assertInvalid("jdbc:trino://localhost:8080?socksProxy=localhost:1080&httpProxy=localhost:8888", "Connection property 'socksProxy' is not allowed");
-        assertInvalid("jdbc:trino://localhost:8080?httpProxy=localhost:8888&socksProxy=localhost:1080", "Connection property 'socksProxy' is not allowed");
+        assertInvalid("jdbc:trino://localhost:8080?socksProxy=localhost:1080&httpProxy=localhost:8888", "Connection property socksProxy cannot be used when httpProxy is set");
+        assertInvalid("jdbc:trino://localhost:8080?httpProxy=localhost:8888&socksProxy=localhost:1080", "Connection property socksProxy cannot be used when httpProxy is set");
```
@wendigo this is the new validation I added that we discussed offline. When adding this, I haven't noticed that in other existing error messages, we quote the property name. I wonder which is better:
* Quotes emphasize the property name;
* But all property names use camel case and can't have spaces in them, so they don't otherwise need quoting
* Not using quotes makes the message a bit more terse

Should I add quotes here or remove them elsewhere to make all message consistent?

@mosabua WDYT? Also, these new validation messages are improving the UX quite a bit, so they might be worth of a release note.

nineinchnick replied:
Please review this new commit https://github.com/trinodb/trino/pull/12587/commits/f762bd1b9c5d13b23ef9daf31319cd6f6fb2c1f0

nineinchnick commented:
Rebased to include the new `KerberosConstrainedDelegation` parameter. I actually found and fixed one issue after the last rebase, where I missed `hostnameInCertificate` in one place. Maybe we're missing a test for this.

wendigo commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1250 @@
+                sslUseSystemTrustStore.orElse(false));
+    }
+
+    public void setupClient(OkHttpClient.Builder builder)
```
in both of the usages we are passing new, fresh OkHttpClient.Builder here.

Can we instead do: `public OkHttpClient.Builder toHttpClientBuilder()` ?

wendigo commented on client/trino-cli/src/main/java/io/trino/cli/QueryRunner.java:
```diff
@@ -59,70 +40,27 @@ public class QueryRunner
-            this.sslSetup = OkHttpUtil::setupInsecureSsl;
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+        try {
+            sslSetup = uri.getSetupSsl();
```
why do we need this sslSetup at all?

Isn't `uri.setupClient(builder);` already doing required SSL initialization?

nineinchnick commented on client/trino-jdbc/src/main/java/io/trino/jdbc/ConnectionProperties.java:
```diff
@@ -288,33 +293,41 @@ public DisableCompression()
+                    validator(
+                            AssumeLiteralUnderscoreInMetadataCallsForNonConformingClients.IS_NOT_ENABLED.or(IS_NOT_ENABLED),
+                            format(
+                                    "Connection property %s cannot be set if %s is enabled",
```
You mean `PropertyName`? These values conflict with connection properties constants defined in this class.

wendigo replied:
Ok

mosabua commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -37,98 +44,165 @@
-    @Option(names = "--source", paramLabel = "<source>", defaultValue = "trino-cli", description = "Name of source making query " + DEFAULT_VALUE)
-    public String source;
+    @PropertyMapping(SOURCE)
+    @Option(names = "--source", paramLabel = "<source>", description = "Name of source making query (default: " + SOURCE_DEFAULT + ")")
```
maybe 

Name of the client to use as source that submits the query. Defaults to trino-cli.

nineinchnick replied:
Done, but I didn't change the default in parentheses, since this needs to be consistent with all other options.

mosabua commented on client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:
```diff
@@ -249,10 +334,121 @@ public ClientSession toClientSession()
+        }
+        catch (RestrictedPropertyException e) {
+            if (e.getPropertyName() == PropertyName.PASSWORD) {
+                throw new IllegalArgumentException("Setting the password in the URL parameter is not allowed, use the `--password` option or the `TRINO_PASSWORD` environment variable");
```
maybe wrap long line?

nineinchnick commented on client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:
```diff
@@ -137,8 +180,10 @@ private TrinoDriverUri(URI uri, Properties driverProperties)
 
         validateConnectionProperties(properties);
 
+        initFromProperties();
```
Done!

nineinchnick commented on client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractConnectionProperty.java:
```diff
@@ -181,6 +184,29 @@ interface Converter<V, T>
 
+    protected interface Validator<T>
+    {
+        Optional<String> validate(T t);
```
Done!

mosabua commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1249 @@
+        useSecureConnection = SSL.getValue(properties).orElse(uri.getScheme().equals("https") || (uri.getScheme().equals("trino") && uri.getPort() == 443));
+        if (!password.orElse("").isEmpty()) {
+            if (!useSecureConnection) {
+                throw new SQLException("Authentication using username/password requires SSL to be enabled");
```
```suggestion
                throw new SQLException("TLS/SSL required for authentication with username and password");
```

mosabua replied:
But can that not be disabled with the insecure config?

nineinchnick replied:
The `--insecure` flag in the CLI is the equivalent of `SSLVerification=NONE`, so it doesn't disable TLS/SSL. In the CLI, enabling TLS is done by specifying the `https` protocol or using the `443` port if there's no protocol.

I changed the error message to `TLS/SSL is required for authentication with username and password`, with the extra `is` before `required`.

wendigo replied:
Insecure means "accept certificate that is not trusted"

nineinchnick replied:
I applied the suggested error.

mosabua commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1249 @@
+            String password = this.password.orElse("");
+            if (!password.isEmpty()) {
+                if (!useSecureConnection) {
+                    throw new SQLException("Authentication using username/password requires SSL to be enabled");
```
```suggestion
                    throw new SQLException("TLS/SSL required for authentication with username and password");
```

mosabua commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1249 @@
+
+            if (kerberosRemoteServiceName.isPresent()) {
+                if (!useSecureConnection) {
+                    throw new SQLException("Authentication using Kerberos requires SSL to be enabled");
```
```suggestion
                    throw new SQLException("TLS/SSL required for Kerberos authentication");
```

mosabua commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1249 @@
+
+            if (accessToken.isPresent()) {
+                if (!useSecureConnection) {
+                    throw new SQLException("Authentication using an access token requires SSL to be enabled");
```
```suggestion
                    throw new SQLException("TLS/SSL required for authentication using an access token");
```

mosabua commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1249 @@
+
+            if (externalAuthentication.orElse(false)) {
+                if (!useSecureConnection) {
+                    throw new SQLException("Authentication using external authorization requires SSL to be enabled");
```
```suggestion
                    throw new SQLException("TLS/SSL required for authentication using external authorization");
```

mosabua commented on client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:
```diff
@@ -0,0 +1,1249 @@
+        for (String queryArg : queryArgs) {
+            List<String> parts = ARG_SPLITTER.splitToList(queryArg);
+            if (parts.size() != 2) {
+                throw new SQLException(format("Connection argument is not valid connection property: '%s'", queryArg));
```
```suggestion
                throw new SQLException(format("Connection argument is not a valid connection property: '%s'", queryArg));
```

mosabua commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriver.java:
```diff
@@ -811,7 +811,7 @@ public void testPropertyAllowed()
                         .buildOrThrow())))
                 .isInstanceOf(SQLException.class)
-                .hasMessage("Connection property 'SSLVerification' is not allowed");
+                .hasMessage("Connection property SSLVerification requires SSL to be enabled");
```
requires TLS/SSL to be enabled

mosabua commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriverAuth.java:
```diff
@@ -319,7 +319,7 @@ public void testFailedCaSslVerificationAlternateHostname()
         assertThatThrownBy(() -> DriverManager.getConnection(url, properties))
                 .isInstanceOf(SQLException.class)
-                .hasMessage("Connection property 'hostnameInCertificate' is not allowed");
+                .hasMessage("Connection property hostnameInCertificate requires SSLVerification to be FULL");
```
```suggestion
                .hasMessage("Connection property hostnameInCertificate requires SSLVerification to be set to FULL");
```

mosabua commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriverAuth.java:
```diff
@@ -340,7 +340,7 @@ public void testFailedNoneSslVerificationAlternateHostname()
         assertThatThrownBy(() -> DriverManager.getConnection(url, properties))
                 .isInstanceOf(SQLException.class)
-                .hasMessage("Connection property 'hostnameInCertificate' is not allowed");
+                .hasMessage("Connection property hostnameInCertificate requires SSLVerification to be FULL");
```
```suggestion
                .hasMessage("Connection property hostnameInCertificate requires SSLVerification to be set to FULL");
```

mosabua commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriverAuth.java:
```diff
@@ -396,7 +396,7 @@ public void testFailedNoneSslVerificationWithSSL()
         assertThatThrownBy(() -> createConnection(ImmutableMap.of("SSLVerification", "NONE")))
                 .isInstanceOf(SQLException.class)
-                .hasMessage("Connection property 'SSLTrustStorePath' is not allowed");
+                .hasMessage("Connection property SSLTrustStorePath cannot be set if SSLVerification is NONE");
```
```suggestion
                .hasMessage("Connection property SSLTrustStorePath cannot be set if SSLVerification is set to NONE");
```

mosabua commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriverUri.java:
```diff
@@ -103,70 +103,70 @@ public void testInvalidUrls()
 
         // ssl verification without ssl
-        assertInvalid("jdbc:trino://localhost:8080?SSLVerification=FULL", "Connection property 'SSLVerification' is not allowed");
+        assertInvalid("jdbc:trino://localhost:8080?SSLVerification=FULL", "Connection property SSLVerification requires SSL to be enabled");
```
```suggestion
        assertInvalid("jdbc:trino://localhost:8080?SSLVerification=FULL", "Connection property SSLVerification requires TLS/SSL to be enabled");
```

mosabua commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriverUri.java:
```diff
@@ -103,70 +103,70 @@ public void testInvalidUrls()
 
         // key store path with ssl verification mode NONE
-        assertInvalid("jdbc:trino://localhost:8080?SSL=true&SSLVerification=NONE&SSLKeyStorePath=keystore.jks", "Connection property 'SSLKeyStorePath' is not allowed");
+        assertInvalid("jdbc:trino://localhost:8080?SSL=true&SSLVerification=NONE&SSLKeyStorePath=keystore.jks", "Connection property SSLKeyStorePath cannot be set if SSLVerification is NONE");
```
```suggestion
        assertInvalid("jdbc:trino://localhost:8080?SSL=true&SSLVerification=NONE&SSLKeyStorePath=keystore.jks", "Connection property SSLKeyStorePath cannot be set if SSLVerification is set to NONE");
```

mosabua commented on client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDriverUri.java:
```diff
@@ -103,70 +103,70 @@ public void testInvalidUrls()
 
         // use system trust store with ssl verification mode NONE
-        assertInvalid("jdbc:trino://localhost:8080?SSLUseSystemTrustStore=true", "Connection property 'SSLUseSystemTrustStore' is not allowed");
+        assertInvalid("jdbc:trino://localhost:8080?SSLUseSystemTrustStore=true", "Connection property SSLUseSystemTrustStore cannot be set if SSLVerification is NONE");
```
```suggestion
        assertInvalid("jdbc:trino://localhost:8080?SSLUseSystemTrustStore=true", "Connection property SSLUseSystemTrustStore cannot be set if SSLVerification is set to NONE");
```

mosabua commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/cli/TestTrinoLdapCli.java:
```diff
@@ -252,7 +252,7 @@ public void shouldFailQueryForLdapWithoutHttps()
         launchTrinoCliWithServerArgument("--execute", SELECT_FROM_NATION);
         assertThat(trimLines(trino.readRemainingErrorLines())).anySatisfy(line ->
-                assertThat(line).contains("Authentication using username/password requires HTTPS to be enabled"));
+                assertThat(line).contains("Authentication using username/password requires SSL to be enabled"));
```
```suggestion
                assertThat(line).contains("TLS/SSL required for authentication with username and password"));
```

mosabua reviewed (changes requested):
Just a whole bunch of minor wording improvements .. essentially though this is good to go. BIG change!

nineinchnick commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/cli/TestTrinoLdapCli.java:
```diff
@@ -229,7 +229,7 @@ public void shouldFailQueryForEmptyUser()
         launchTrinoCliWithServerArgument("--execute", SELECT_FROM_NATION);
         assertThat(trimLines(trino.readRemainingErrorLines())).anySatisfy(line ->
-                assertThat(line).contains("Malformed credentials: user is empty"));
+                assertThat(line).contains("Username must be specified along with password"));
```
I changed this to `Both username and password must be specified` to emphasize that only one of them is missing.

wendigo commented:
There is related failure in suite-ldap:

```
tests               | Expecting throwable message:
tests               |   "Authentication using username/password requires SSL to be enabled"
tests               | to contain:
tests               |   "TLS/SSL is required for authentication with username and password"
tests               | but did not.
```

wendigo commented:
@nineinchnick can you update release notes:

```
Trino CLI now accepts a single URL with parameters as an alternative to passing command line arguments.
```

wendigo commented on testing/trino-product-tests/src/main/resources/tempto-configuration.yaml:
```diff
@@ -46,7 +46,7 @@ databases:
     jdbc_url: jdbc:trino://${databases.presto.host}:${databases.presto.port}/hive/tpcds
     jdbc_user: hive
-    jdbc_password: "***empty***"
+    jdbc_password: ""
```
<3

mosabua reviewed (approved):
All good now. Lets ship this!

wendigo commented:
All tests have passed. I have a small refactor around creating http client used by the jdbc and client but I’ll follow up in a separate PR. Merging this as-is for now. Thanks @nineinchnick for improving CLI and making it consistent with JDBC driver.

nineinchnick commented:
BTW there's one last improvement to be done in JDBC error messages, when the value is being converted, for example from a string to a boolean, the error now looks like:
> Connection property SSL value is invalid: 2

But we could say how the expected values look like, for example:
> Connection property SSL value '2' is invalid, expected 'true' or 'false'

I'll follow up with another small PR.
